### DISCUSSION TITLE
How to delete an instrument's notes using MIDI messages?

### DISCUSSION THREAD

**Discussion by jklerch** [hydrogen-music/hydrogen#1948] [Q&A]
Hello,

Recently, I found out how to use Hydrogen during live performances with a MIDI controller: http://hydrogen-music.org/documentation/manual/manual_en_chunked/ch16.html Almost all functions I need are found in hydrogen. This is awesome!

My workflow is the following: I have one pattern looping and record notes from MIDI input during the loop. These notes are then played back immediately when the loop starts over.

However, I did not find out how to **delete the notes of an instrument using a MIDI controller**. This would be important if I have some beat running right now, but I want to change say the pattern of the kick. I could easily just right click and say "delete notes". However, during a live performance, this is not really practical. It would be better if there would be the option to map this function to some button on my MIDI controller. 

Is there a way to do that I maybe did not find yet? Otherwise, would it be possible to implement the possibility to map that?

**Comment by theGreatWhiteShark** [comment]
Hey @jklerch ,

> However, I did not find out how to delete the notes of an instrument using a MIDI controller. This would be important if I have some beat running right now, but I want to change say the pattern of the kick. I could easily just right click and say "delete notes". However, during a live performance, this is not really practical. It would be better if there would be the option to map this function to some button on my MIDI controller.
>
> Is there a way to do that I maybe did not find yet? Otherwise, would it be possible to implement the possibility to map that?

Nice suggestion! No, we do not have such a MIDI command yet. But it sounds like a very good idea to add one.

I would suggest to call it `CLEAR_INSTRUMENT` and make it work similar to `SELECT_INSTRUMENT` while always acting on the currently selected pattern.

In addition, I would also add a `CLEAR_PATTERN` to delete all notes of the currently selected pattern. It's cumbersome to clear every single instrument individually and as there is no way to add a pattern via MIDI one could quickly run out of patterns by just switching to a blank one.

What do you think?

**Reply by jklerch** [reply]
Thank you for your fast response. I do not have much experience with MIDI in Hydrogen, so I trust you that the proposed solution is probably the one that is most consistent with everything else.

Another thought: as I explained, it is nice for me to delete the whole pattern with just one click (exactly, like you proposed. I like it exactly like that). However, other sequencers follow the approach that you say "during the loop running: let's press a button that somehow corresponds to `delete notes of instrument x`, but that button does not delete all notes at one click, but it deletes those notes of that instrument that are played during the button is pressed down". This is an alternative approach how to delete notes. It allows you to be more selective which exact note to delete. I think it is important to have that in mind, because there might be other users with other workflows. 

However, I just wanted to mention this for the sake of completeness. To me it would be totally fine to have the `CLEAR_INSTRUMENT` functionality. To keep things simple. If people have capacities to care about more fancy approaches that could be implemented as a separate functionality in my opinion.

**Reply by theGreatWhiteShark** [reply]
> However, other sequencers follow the approach that you say "during the loop running: let's press a button that somehow corresponds to delete notes of instrument x, but that button does not delete all notes at one click, but it deletes those notes of that instrument that are played during the button is pressed down". This is an alternative approach how to delete notes. It allows you to be more selective which exact note to delete. I think it is important to have that in mind, because there might be other users with other workflows.

It would be interesting to know how they accomplish this. Recording a note is done based on `NOTE_ON` events, which feature a velocity - how hard you hit - and a pitch. The former is directly translated to the note velocity in Hydrogen and the latter is mapped to a specific instrument. Other MIDI event types, like `CC`, `PC`, or `MMC_*`, do feature a value - similar to the velocity - but the other parameter - if any - is hard coded to identify the particular button. Using something like a knob to delete per-instrument notes would be really cumbersome as it only sends *changes* (in the instrument).

Or maybe they use `NOTE_OFF` messages to erase a note. But that's not possible in Hydrogen. We already have a feature - when disabling "ignore Note Off" in Preferences > Midi - to record those events (to restrict the duration of a note to a certain length).

Or maybe they have a MIDI event for toggling between recording and deleting notes.

**Reply by jklerch** [reply]
I know that the Roland SP-404 SX Sampler does it like this in its sequencer functionality. My source is [this video at the timecode provided](https://youtu.be/daYrbCaUaxc?feature=shared&t=262). From that, I assume you can implement it using the last approach you mentioned.

**Reply by theGreatWhiteShark** [reply]
> From that, I assume you can implement it using the last approach you mentioned.

That approach is really growing on me. We could have a `RECORD/DELETE_TOGGLE` or `DELETE_NOTES_TOGGLE`.

One question popping up right away is whether to playback the sound corresponding to the note. Traditionally all incoming NOTE_ON events - except those being discarded after triggering an action - will attempt to trigger a sound. Only in a second step the note might be recorded in case the corresponding button was pressed. For the deletion I would do the same. Especially since there is no visual representation of this state (yet), which could lead to awkward situations where the user might think MIDI input is broken.

Is this behavior something you would expect?

**Reply by jklerch** [reply]
I did not quite understand the behavior you are describing. Nevertheless, here is what I would expect:

To delete a note:
1. press button to toggle delete mode
2. press and hold the button corresponding to the instrument where notes should be deleted
    - thereby, if a previously recorded note_on event would be played while the button is pressed, I would assume it not to be played at all, but deleted instead. So I would assume it to behave just as with the SP-404 SX (see video I sent in the message above)

(btw: thank you very much for your engagement of implementing this feature and consideration of user experience!)

**Reply by theGreatWhiteShark** [reply]
> I did not quite understand the behavior you are describing.

Yeah. That's what I thought. Some veteran Hydrogen users might not be surprised by this behavior but prolonging the "old ways" would not yield the clean and intuitive UX we strive for.

> Nevertheless, here is what I would expect:
> 
> To delete a note:
>
>  1.  press button to toggle delete mode
>  2. press and hold the button corresponding to the instrument where notes should be deleted
        thereby, if a previously recorded note_on event would be played while the button is pressed, I would assume it not to be played at all, but deleted instead. So I would assume it to behave just as with the SP-404 SX (see video I sent in the message above)

Yes. This sounds better. However, it will also take significantly longer. Without an equivalent of the "DEL" button of the SP-404 SX representing the state of the toggle and allowing it to be altered via the UI, this feature will only be a stub. But this will only happen for 1.3 when a number of other are already implemented.

By the way, which OS are you using?

**Reply by jklerch** [reply]
Do you mean it will take longer to develop or do you mean the speed of the operation might be so slow that it will be inconvenient to use this feature?

If it will just take time to implement it, I would propose to go with this "better" option, even if it takes more time. In parallel (and if there is capacity to implement it) one could add another feature that just deletes the whole instrument note sequence with one press of a button (as we discussed earlier in this thread).

I am using Arch Linux. EndeavourOS to be precise.

**Reply by theGreatWhiteShark** [reply]
> Do you mean it will take longer to develop or do you mean the speed of the operation might be so slow that it will be inconvenient to use this feature?

Longer to develop.

> If it will just take time to implement it, I would propose to go with this "better" option, even if it takes more time. In parallel (and if there is capacity to implement it) one could add another feature that just deletes the whole instrument note sequence with one press of a button (as we discussed earlier in this thread).

My thoughts precisely.

> I am using Arch Linux. EndeavourOS to be precise.

Thanks. Could you test in [this version](https://ci.appveyor.com/api/buildjobs/7r9938bnxdoyy560/artifacts/Hydrogen-1.2.3-19-g8358aee1b-x86_64.AppImage) whether `CLEAR_PATTERN` and `CLEAR_INSTRUMENT` work as expected?

**Reply by jklerch** [reply]
I have another question (but I suppose direct messages are not possible on GitHub), so I ask it here and I hope it is okay:

Will you be at any event by the CCC in Germany this year? Maybe we could meet there and exchange some further information about the project.

**Reply by theGreatWhiteShark** [reply]
Hmm.. Nothing planned yet. Maybe I'll attend the Datenspuren in Dresden. But you could also drop a message in case you are visiting a hackspace in Leipzig.

**Reply by jklerch** [reply]
Allright.

The following things come to mind:
- `CLEAR_PATTERN` works as expected. However, it is not possible to me undoing this action using Ctrl+Z
- `CLEAR_INSTRUMENT` does not work as expected. I expect it to work in the following way: set `SELECT_INSTRUMENT` to the instrument I want to have deleted; execute `DELETE_INSTRUMENT` -> instrument notes get deleted. What actually happens is: when executing `DELETE_INSTRUMENT`, the last instrument (in my case, instrument 8) gets selected and deleted. No matter which instrument I chose before.

After selecting instrument 2 (`SELECT_INSTRUMENT`):
![image](https://github.com/hydrogen-music/hydrogen/assets/86262129/c3a6d29a-5976-4a0e-a11a-fbd5903d97e0)

After executing `DELETE_INSTRUMENT`:
![image](https://github.com/hydrogen-music/hydrogen/assets/86262129/f65b0a5d-6b41-4b17-912b-cef50f72cb7e)

**Reply by theGreatWhiteShark** [reply]
> CLEAR_INSTRUMENT does not work as expected. I expect it to work in the following way: set SELECT_INSTRUMENT to the instrument I want to have deleted; execute DELETE_INSTRUMENT -> instrument notes get deleted. What actually happens is: when executing DELETE_INSTRUMENT, the last instrument (in my case, instrument 8) gets selected and deleted. No matter which instrument I chose before.

Hmm... is there a use case for separating the steps of selecting an instrument and deleting all its notes? I drafted `CLEAR_INSTRUMENT` to work similar to `SELECT_INSTRUMENT`. It reads an instrument number from the MIDI input value, selects the corresponding instrument, and deletes its notes. Hence the similarity in naming.

But to answer my own question: when assigning the action to a knob with current value of `0`, clearing instrument `10` would result in clearing all instruments in between. That's not convenient....

How would you use this feature?

**Reply by jklerch** [reply]
At first: My personal opinion might not be general and not be applicable to everyone's workflow. (I suppose you also assume that, but wanted to be clear here).

If you implemented it the way I understand from your message, then one would have to assign `number_of_instruments x 2` many "buttons": `number_of_instruments` many to set a note and again, `number_of_instruments` many to delete a note, did I understand that correctly?

If this is the case: You would need many "buttons" on your controller, but you might not have that many ones... So to me it would just be more feasible to have _one_ button for "delete instrument". This button would then delete the currently selected instrument. One could have another poti assigned for selecting an instrument using `SELECT_INSTRUMENT`.

This would be the way I would use the solution that was implemented by you. However, I do not know how that complies with all the other functionalities and if it makes things more chaotic. Also, I am new to Hydrogen MIDI stuff, so if there is another "Hydrogen way" to do things (and which is also for some reason "the best possible" way under the given configuration. I.e. better than my proposal in the above paragraph), then I would be ready to learn that as long as it is feasible with the controller I have available.

**Reply by theGreatWhiteShark** [reply]
> At first: My personal opinion might not be general and not be applicable to everyone's workflow. (I suppose you also assume that, but wanted to be clear here).

Sure. But I never used Hydrogen live. Maybe I miss some entirely obvious points.

> If you implemented it the way I understand from your message, then one would have to assign number_of_instruments x 2 many "buttons": number_of_instruments many to set a note and again, number_of_instruments many to delete a note, did I understand that correctly?

It was more an action to assign on a poti. But scratch that. In terms of MIDI it is not really helpful.

I would drop the `CLEAR_INSTRUMENT` MIDI action but keep the OSC counterpart (because there it is much more easy specify particular instrument). Instead, I would introduce `CLEAR_SELECTED_INSTRUMENT` which takes no input and can be assigned to a button.

https://ci.appveyor.com/api/buildjobs/ak396wmsa47n5ykk/artifacts/Hydrogen-1.2.3-25-g402f09b3c-x86_64.AppImage

**Reply by theGreatWhiteShark** [reply]
Hey @jklerch ,

I merged the state described above and it will be included in the 1.2.4 release of Hydrogen. In addition, I split of the record/delete toggle into a separate issue (#1965). I think we are done here, aren't we?
